### ISSUE TITLE
plugin/metrics: add host/server ID

### ISSUE THREAD

**Issue by miekg** [coredns/coredns#1189]
This is probably a good idea for us as well:
https://github.com/miekg/caddy-prometheus/pull/32

**Comment by johnbelamaric** [comment]
Something like this could be good - in the `telemetry` plugin discussion we have talked about naming deployments, something like this may make it unnecessary. @chrisohaver

**Comment by greenpau** [comment]
> Something like this could be good - in the telemetry plugin discussion

@johnbelamaric , is there such a plugin?

> This is probably a good idea for us as well:

>> hostname local-1

@miekg , the implementation in [caddy-prometheus PR](https://github.com/miekg/caddy-prometheus/pull/32/commits/37c0c9da8626693459f089f3395445d51ae735a0#diff-8b80bab7c379f59fe83cb4ee2d015ea7R16) creates a single label "hostname". What do you think about having multiple labels?

from:

```
. {
    prometheus localhost:{$PORT}
}
```

to:

```
. {
    prometheus localhost:5555 hostname "local-1" version "1.0.0" contact "Miek Gieben"
}
```

**Comment by miekg** [comment]
re: telemetry plugin: I thought infoblox had some internal prototyping thing going on.

What's the benefit of multiple labels?

**Comment by johnbelamaric** [comment]
Yes, @chrisohaver is working on it.


On Dec 15, 2017, at 5:47 AM, Miek Gieben <notifications@github.com<mailto:notifications@github.com>> wrote:


re: telemetry plugin: I thought infoblox had some internal prototyping thing going on.

What's the benefit of multiple labels?

**Comment by chrisohaver** [comment]
Multiple labels can help with reporting: For example, they can enable a single prometheus instance to report on separate deployments of coredns,  each deployment potentially containing multiple coredns instances. e.g. reports grouped by deployment.  More labels could allow reports on groups of deployments

For the same reason, the telemetry prototype (early WIP) adds two metric labels to delineate deployments and instances within an organization. However, it does this when pushing metrics to the reporting service, it doesn't add them to the original metrics (which, if we do, will make this step unnecessary).

**Comment by miekg** [comment]
There might be a case of adding some k8s specific labels, but only if you can't do this from the prometheus side of things

**Comment by greenpau** [comment]
@miekg , @chrisohaver, @johnbelamaric , I would like to get working on this one.

What do you think about having the following syntax for Prometheus metrics?

```
. {
    prometheus {
        listen HOST:PORT
        label NAME VALUE
    }
}
```

That would allow to do the following:

```
. {
    prometheus {
        listen localhost:5555
        label hostname "host-a1"
        label version "1.0.0"
        label contact "Paul Greenberg"
    }
}
```

**Comment by miekg** [comment]
[ Quoting <notifications@github.com> in "Re: [coredns/coredns] plugin/metric..." ]
>@miekg , @chrisohaver, @johnbelamaric , I would like to get working on this one.
>
>What do you think about having the following syntax for Prometheus metrics?

Why can't you do this in prometheus? If you have one stable label (a `host id` 
for instance) you can relabel and what not?

**Comment by johnbelamaric** [comment]
@miekg I think that would mean customizing the prometheus config for each instance, even when they come and go in, say, a k8s cluster. With this those instances can be auto-discovered with prometheus service discovery. maybe there is a way in the prometheus config to copy labels out of k8s? But if you aren't using k8s then you are stuck in that case.

**Comment by greenpau** [comment]
> I think that would mean customizing the prometheus config for each instance, even when they come and go in, say, a k8s cluster.

@johnbelamaric , 👍

**Comment by miekg** [comment]
fair enough, but how indicative is the above example as that doesn't make the use case clear

**Comment by miekg** [comment]
See #1492 
hopefully fixed in 1.1.2

**Comment by miekg** [comment]
fixed with per-server metrics
